File: python/剑指30包含min函数的栈.py
class MinStack:

    def __init__(self):
        
        self.st=[]
        # 辅助栈
        self.as_st= [float('inf')]

    def push(self, x: int) -> None:
        self.st.append(x)
        self.as_st.append(min(self.as_st[-1],x))

    def pop(self) -> None:
        if(len(self.st)):
            self.st.pop()
            self.as_st.pop()

    def top(self) -> int:
        if(len(self.st)):
            return self.st[-1]

    def min(self) -> int:
        if(len(self.st)):
            return self.as_st[-1]

File: csharp/剑指30包含min函数的栈.cs
public class MinStack
{

    /** initialize your data structure here. */
    Stack<int> st = new();
    Stack<int> as_st = new();
    public MinStack()
    {
        as_st.Push(int.MaxValue);
    }

    public void Push(int x)
    {
        st.Push(x);

        as_st.Push(Math.Min(x, as_st.Peek()));
            
    }

    public void Pop()
    {
        st.TryPop(out _);
        as_st.TryPop(out _);
    }

    public int? Top()
    {
        int res;
        if(st.TryPeek(out res))
            return res;
        return null;
    }

    public int? Min()
    {
        int res;
        if(as_st.TryPeek(out res))
            return res;
        return null;
    }
}
